src/meili/routing.cc

```diff
@@ -524,43 +524,91 @@ find_shortest_path(baldr::GraphReader& reader,
       }
 
       // Expand origin: add segments from origin to destinations ahead
-      // at the same edge to the queue
-      if (dest == origin_idx) {
+      // at the same edge as well as at the opposite edge to the queue
+      if (destination_idx == origin_idx) {
         for (const auto& origin_edge : destinations[origin_idx].edges) {
           // The tile will be guaranteed to be directededge's tile in this loop
-          const baldr::GraphTile* tile = nullptr;
-          const auto directededge = reader.directededge(origin_edge.id, tile);
+          const baldr::GraphTile* start_tile = nullptr;
+          const auto directed_edge = reader.directededge(origin_edge.id, start_tile);
+
+          baldr::GraphId opposite_edgeid = reader.GetOpposingEdgeId(origin_edge.id);
+          const baldr::GraphTile* end_tile = directed_edge->leaves_tile()
+                                                 ? reader.GetGraphTile(directed_edge->endnode())
+                                                 : start_tile;
+          const baldr::DirectedEdge* opposite_edge = reader.directededge(opposite_edgeid, end_tile);
```

Comment:
i am pretty sure we have alraedy a helper function that does all of this in one call to graphreader.